### PR TITLE
C++: Bring 'swap' models together.

### PR DESCRIPTION
**Pull request by geoffw0**
Bring the various `swap` member function models together into one place, merge them, then generalize.  The last commit (generalizing them) is perhaps controversial as it means we no longer control the model so tightly - but it's pretty well constrained and I expect results to be good.  Discussion encouraged.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll [hunk 1]
```diff
@@ -2,7 +2,10 @@ import semmle.code.cpp.models.interfaces.DataFlow
 import semmle.code.cpp.models.interfaces.Taint
 
 /**
- * The standard function `swap`.
+ * The standard function `swap`. A use of `swap` looks like this:
+ * ```
+ * std::swap(obj1, obj2)
+ * ```
  */
 private class Swap extends DataFlowFunction {
```
Comment (MathiasVP) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll:10]: We might as well implement `AliasFunction` while we're at it, right? (Doesn't have to be in this PR, of course)

File: cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll [hunk 2]
```diff
@@ -15,3 +18,27 @@ private class Swap extends DataFlowFunction {
...
+ * A `swap` member function that is used as follows:
+ * ```
+ * obj1.swap(obj2)
+ * ```
+ */
+private class MemberSwap extends TaintFunction, MemberFunction {
+  MemberSwap() {
+    this.hasName("swap") and
+    this.getNumberOfParameters() = 1 and
+    this.getParameter(0).getType() instanceof ReferenceType and
```
Comment (MathiasVP) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll:32]: This line seems unnecessary given the postfix cast just below it.

**Review by MathiasVP — COMMENTED** [review]
I think this is a good change. I can't imagine that we want to distinguish these different swap implementation.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll:11]: Done.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll]: Good point - removed.

**Comment by geoffw0** [comment]
Comments addressed and merge conflict fixed.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll [hunk 3]
```diff
@@ -15,3 +19,32 @@ private class Swap extends DataFlowFunction {
...
+    input.isQualifierObject() and
+    output.isParameterDeref(0)
+    or
+    input.isParameterDeref(0) and
+    output.isQualifierObject()
+  }
+
+  override predicate parameterNeverEscapes(int index) { none() }
+
+  override predicate parameterEscapesOnlyViaReturn(int index) { index = 0 }
```
Comment (MathiasVP) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll:47]: Does this combination really make sense? How can a parameter never escape, but also escape via returns? Should `parameterNeverEscapes` specify that both arguments never escape?

Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll:47]: `parameterNeverEscapes` has an empty implementation (`AliasFunction` is set up in such a way that there has to be a definition here).

Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll:47]: Oh! The `Never` part of the `parameterNeverEscapes` predicate had me confused. You're right, all is good!

**Review by MathiasVP — APPROVED** [review]
LGTM!

File: cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll (hunk 3 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Swap.qll:47]: Yeah, I find the `AliasFunction` model a bit confusing and always seem to have to read everything twice.

**Pull request merged**